docs: Refactor ebpf installation (#2849)
grafana/pyroscope: docs/sources/configure-client/grafana-agent/ebpf/_index.md

---
title: "Profiling with eBPF with Grafana Agent"
menuTitle: "Profiling with eBPF"
description: "Learn about using eBPF for continuous profiling for performance optimization."
weight: 20
aliases:
  - /docs/phlare/latest/configure-client/language-sdks/ebpf/
  - /docs/pyroscope/next/configure-client/language-sdks/ebpf
---

# Profiling with eBPF with Grafana Agent

> <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b0/EBPF_logo.png/240px-EBPF_logo.png" width="100px;" alt="eBPF"/>

Reviewer: I'm not sure about using the logo here. It's probably okay to use the image, but I'm not sure if we should upload it to the Grafana servers or call it from the wikimedia site directly. I'll find out and let you know.

Reviewer: Okay, looks like we can use the logo, we just have to upload it to our server. I'll do that and update the URL.

Reviewer: ```suggestion
<img src="/media/docs/pyroscope/ebpf_logo_color_on_white.png" width="100px;" alt="eBPF"/>
```

eBPF is an advanced technology embedded into the Linux kernel. It stands for enhanced [Berkeley Packet Filter](https://en.wikipedia.org/wiki/EBPF) and revolutionizes the capability to run sandboxed code safely within the kernel space. This technology serves multiple use cases, such as networking, security, and performance monitoring without the need to alter kernel code or load additional modules.

> ## Benefits and Tradeoffs of using eBPF for continuous profiling

Reviewer: ```suggestion
## Benefits and tradeoffs of using eBPF for continuous profiling
```

> When it comes to application profiling, eBPF shines due to its high efficiency and minimal performance overhead. It enables the dynamic insertion of powerful monitoring code into live production systems. By leveraging eBPF, developers can gain insights into application behavior, track resource usage, and detect bottlenecks in a way that traditional profiling tools cannot match. eBPF's low overhead and fine-grained data collection make it an ideal choice for continuous, real-time profiling in performance-sensitive environments.
> 
> However, eBPF is not a silver bullet. It has some limitations that make it unsuitable for certain use cases:
> 
> - For example, eBPF is not a good fit for profiling applications that are not written in a supported language.
> - It cannot be used to profile applications that are not running on Linux.
> - It does not support all profile types such as memory and contention/lock profiling.
> - eBPF requires root access to the host machine, which can be a problem in some environments.

Reviewer: Simplifying some of the wording and putting some of the longer sentences on to new lines. 
```suggestion
When it comes to application profiling, eBPF shines due to its high efficiency and minimal performance overhead.
eBPF enables the dynamic insertion of powerful monitoring code into live production systems.
By leveraging eBPF, developers can gain insights into application behavior, track resource usage, and detect bottlenecks in a way that traditional profiling tools cannot match.
eBPF's low overhead and fine-grained data collection make it an ideal choice for continuous, real-time profiling in performance-sensitive environments.

However, eBPF has some limitations that make it unsuitable for certain use cases:

- It isn't a good fit for profiling applications that are not written in a supported language.
- It can't be used to profile applications that are not running on Linux.
- It does not support all profile types such as memory and contention/lock profiling.
- eBPF requires root access to the host machine, which can be a problem in some environments.
```

---
title: "eBPF"
menuTitle: "eBPF"
description: "Understanding eBPF: Continuous Profiling for Performance Optimization"
weight: 20
aliases:
  - /docs/phlare/latest/configure-client/language-sdks/ebpf/
  - /docs/pyroscope/next/configure-client/language-sdks/ebpf
---

> # Profiling with <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b0/EBPF_logo.png/240px-EBPF_logo.png" width="100px;" alt="eBPF"/>

Reviewer: We don't tend to use images in the headings. Also the page title and title in the Frontmatter should match. I won't start a review, but I'll go ahead and provide a suggestion to correct this.

> title: "eBPF"
> menuTitle: "eBPF"
> description: "Understanding eBPF: Continuous Profiling for Performance Optimization"
> weight: 20
> aliases:
>   - /docs/phlare/latest/configure-client/language-sdks/ebpf/
>   - /docs/pyroscope/next/configure-client/language-sdks/ebpf
> ---
> 
> 
> # Profiling with <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b0/EBPF_logo.png/240px-EBPF_logo.png" width="100px;" alt="eBPF"/>

Reviewer: ```suggestion
title: "Profiling with eBPF with Grafana Agent"
menuTitle: "Profiling with eBPF"
description: "Learn about using eBPF for continuous profiling for performance optimization."
weight: 20
aliases:
  - /docs/phlare/latest/configure-client/language-sdks/ebpf/
  - /docs/pyroscope/next/configure-client/language-sdks/ebpf
---

# Profiling with eBPF
```

## Benefits and Tradeoffs of using eBPF for continuous profiling

When it comes to application profiling, eBPF shines due to its high efficiency and minimal performance overhead. It enables the dynamic insertion of powerful monitoring code into live production systems. By leveraging eBPF, developers can gain insights into application behavior, track resource usage, and detect bottlenecks in a way that traditional profiling tools cannot match. eBPF's low overhead and fine-grained data collection make it an ideal choice for continuous, real-time profiling in performance-sensitive environments.

However, eBPF is not a silver bullet. It has some limitations that make it unsuitable for certain use cases:

- For example, eBPF is not a good fit for profiling applications that are not written in a supported language.

> - It also cannot be used to profile applications that are not running on Linux.

Reviewer 2: ```suggestion
- It cannot be used to profile applications that are not running on Linux.
```

- It does not support all profile types such as memory and contention/lock profiling.

> - Finally, setting up eBPF requires root access to the host machine, which can be a problem in some environments.

Reviewer 2: ```suggestion
- eBPF requires root access to the host machine, which can be a problem in some environments.
```

> We recommend an hybrid approach for better results: eBPF to profile the node and specific language instrumentation per application.

Reviewer 2: Id just remove this.. in reality I'd say we recommend eBPF if you're using a supported language and sdks if not

> > We added a blog post "[The pros and cons of eBPF profiling](https://pyroscope.io/blog/ebpf-profiling-pros-cons)" which more deeply

Reviewer 2: I'd also remove this... I think now our eBPF integration has become more advanced and this blog post no longer accurately reflects how we think about our eBPF integration

explores this topic and provides some examples of eBPF profiles. If you're interested in some of the more granular details you can find them there!

> ## Grafana Agent

Reviewer 2: ```suggestion
## eBPF via the Grafana Agent
```

The Grafana Agent is a lightweight, all-in-one agent that can collect, transform, and ship observability data. For profiling, the Grafana Agent can be configured to collect eBPF profiles and send them to Pyroscope.

> This section contains instructions for installing and configuring the Grafana Agent to collect eBPF profiles. For more information about the Grafana Agent itself, see the [Grafana Agent documentation](https://grafana.com/docs/agent/latest/flow/).

Reviewer: ```suggestion
This section contains instructions for installing and configuring the Grafana Agent to collect eBPF profiles. For more information about the Grafana Agent itself, see the [Grafana Agent documentation](/docs/agent/latest/flow/).
```
